### ISSUE TITLE
Unexpected alignment when using boolean `b1`

### ISSUE THREAD

**Issue by StefanD986** [kaitai-io/kaitai_struct#642]
I discovered some unexpected result when trying to parse data that has booleans. Not sure whether it **is** a bug, but it feels like one, so putting it up here for discussion. 

# Steps to reproduce:
Given the attached grammar (see below)
1. Parse the following data: `bytearray([0x69, 0x01, 0x01, 0x5F, 0x00])` (see gist at the end for a python test fixture)

## Expected behaviour
Parsing the data should produce an object structure that essentially contains the following data
```python
[
    {'tag': 105, 'multi_byte_len_flag': False, 'length': 1, 'body': True},
    {'tag': 95, 'multi_byte_len_flag': False, 'length': 0, 'body': None}
]
```

## Actual behaviour
Parsing the data throws the folloowing error: `EOFError: requested 2 bytes, but got only 0 bytes`

# Further notes
Upon inspection of the generated code it appears as if there might be a alignment instruction missing after reading the body data of elements with `tag==105` 

https://gist.github.com/StefanD986/e8d7ca9814a070d7bba74044a7d28ae7#file-tlv_debugging-py-L45

I am following the kaitai documentation, that suggests to use the 1 bit `b1` type for booleans. However I also argue that only evaluating the lowest bit to decide whether the value is `True` or `False` is not enough, since many languages interprete anyting non-zero as *truthy*.

# Attachments

```yaml
meta:
  id: tlv_debugging
  file-extension: hex
  endian: le
  encoding: ascii
seq:
  - id: file_hdr0
    type: header_obj
    repeat: until
    repeat-until: _.length == 0
types:
  header_obj:
    seq:
      - id: tag
        type: u1
      - id: multi_byte_len_flag
        type: b1
      - id: length
        type: b7
      - id: body
        size: length
        type:
          switch-on: tag
          cases:
            95:  tb
            105: b1
    types:
      tb: 
        seq:
          - id: no_value
            size: 0

```

Minmal example code can also be found here: https://gist.github.com/StefanD986/e8d7ca9814a070d7bba74044a7d28ae7

**Comment by generalmimon** [comment]
> I am following the kaitai documentation, that suggests to use the 1 bit `b1` type for booleans. However I also argue that only evaluating the lowest bit to decide whether the value is `True` or `False` is not enough, since many languages interprete anyting non-zero as _truthy_.

Kaitai Struct documentation doesn't suggest using `b1` for every field that should be interpreted as boolean. It just says that if you declare a field of type `b1`, its value won't be either 0 or 1, but booleans `True` or `False`. This is because probably the most common use case for `b1` is parsing flag bitfields. That is, for example some byte which numerical value is not important, but each of its 8 bits has its own meaning. These are usually some switches which turn certain features or settings on or off. The Kaitai Struct automatically converts the numerical value of the bit to boolean, because it's convenient in this case.

But you are describing another problem, you want to read a byte and interpret it as true/false **based on its numerical value** (I suppose "0 => `false`, everything else => `true`" is what you want). This should be achieved by declaring a `u1` (or `s1`) field and using `(field != 0)` everywhere you want to use its value.

And to the bug you describe: I realized that it might be very hard to resolve whether to insert alignment instruction or not. There are so many cases that need to be solved separately, such as primitive `b1` followed by `u1` hidden in the sequence of a user type, `b1` hidden in a user type followed by primitive `u1`, `b1` hidden in a user type followed by `u1` in a user type too, etc. Not even counting the number of possibilities when combining user types with own substream/without it, with `switch` or without it, with `repeat` or without, conditional fields with `if` etc. I think all the hard work isn't worth it, especially when the problem has relatively simple workaround:

```yaml
meta:
  id: bits_bytes_user_switch_aligned
seq:
  - id: problematic_field
    type:
      switch-on: 42
      cases:
        42: b1
        _: u1
    # want to always align to byte here? no problem!
  - type: b0
    # there will be the desired align_to_byte() call
  - size: 0
```

And I also think that even if we got all that work done, it may be very confusing for the programmer if the compiler decided to align to byte somewhere where it wasn't supposed to. That would be much worse than the current situation, because there isn't a way to revert the alignment that has been already done without altering the generated code.

Personally I would leave it as is, but the documentation should probably mention this issue.

**Comment by KOLANICH** [comment]
>`95: tb`

probably can be removed as `tb` is empty.

>`size: 0`

size is 0 bytes, which assumes byte alignment.

See also  #12 and #576.

**Comment by generalmimon** [comment]
@KOLANICH Have you read the comments? It seems like you randomly pick 2 lines on this page and do some code review.

> probably can be removed as `tb` is empty.

Yes, the `tb` field probably is not necessary, but that's not so important. It demonstrates the discussed issue well. Speaking of which, `file-extension: hex`, `endian: le` and `encoding: ascii` are also not needed.

> size is 0 bytes, which assumes byte alignment.

Are you responding to the first https://github.com/kaitai-io/kaitai_struct/issues/642#issue-517145790 by @StefanD986 or my comment? I'll try to analyze both options:

1. In this case, no, the `no_value` field is not preceded by any bit field in the `tb/seq`. That's the point of this issue. It's a question if the compiler should analyze whether the stream position may be byte-unaligned even across type boundaries. Right now, it works only in the same `seq`uence.

2. Yeah, that's the point of my trick, doing `type: b0` doesn't read anything, but persuades compiler that the current stream position is unaligned, the following `size: 0` forces compiler to call `alignToByte`.

**Comment by KOLANICH** [comment]
> @KOLANICH Have you read the comments? It seems like you randomly pick 2 lines on this page and do some code review.

I have. In fact I have misinterpreted the question (I usually had issues when KSC generates the code aligning though the alignment is not needed, so I had to patch the generated source).

>Upon inspection of the generated code it appears as if there might be a alignment instruction missing after reading the body data of elements with tag==105

It generates the code, ignoring the `length` for that branch though it must use it:

```python
            if _on == 95:
                self._raw_body = self._io.read_bytes(self.length)  #length is used
                io = KaitaiStream(BytesIO(self._raw_body))
                self.body = self._root.HeaderObj.Tb(io, self, self._root)
            elif _on == 105:
                self.body = self._io.read_bits_int(1) != 0 # length is ignored!
            else:
                self.body = self._io.read_bytes(self.length)
```

Instead on every `size` KSC should create a substream, IMHO. Probably a one without copying, just a substream using the same underlying stream instead of using `BytesIO`  (not yet implemented).

**Comment by mewmew** [comment]
For what it's worth, I was also hit by this issue today. As a new and very excited user of Kaitai Struct, I jumped straight in, and read enough of the docs to be dangerous :) As such, I interpreted `b1` as meaning *boolean*.

Well, that was until I used the Kaitai Struct IDE on a sample input and realized `b1` reads *one bit* of the stream, not one byte. To a new user, this was confusion at first since `u1`, `s1` both read 1 byte. However, it becomes clear once you start to think of the `b` prefix in terms of number of bits in a bit stream.

As *boolean* is a rather common field type of various binary formats, it would seem useful to users of Kaitai Struct if a `bool` pseudo-type was defined, similar to how `str` is treated for strings.

I propose the introduction of `bool` as a pseudo-type for *boolean* values and furthermore propose that the `size` field may be used in combination with the `bool` type to specify the size (in number of bytes) of the boolean field.

Previously, a user would write `u4` for 32-bit booleans and `u1` for 8-bit booleans (both of which are common in C code). The single bit boolean `b1` is common in bitstream formats (e.g. FLAC), but byte-sized and dword-sized boolean values are also quite common.

So, to sum up, as a new user of Kaitai Struct, this is about the only issue I ran into. Everything else has been a breeze and especially since I started to use the Kaitai Struct IDE, writing format specifications became quite a fun an exhilarating experience.

Keep up the awesome work!

And happy reversing!

Cheers,
Robin

**Comment by generalmimon** [comment]
Duplicate of https://github.com/kaitai-io/kaitai_struct/issues/1070
